### PR TITLE
[13.0] mig l10n_br_base

### PR DESCRIPTION
**Pull request by rvalyi**
Not a module Akretion is interrested in maintaining actively on v13, but migrating to v13 before migrating to v14 was easier than migrating straight, so if anyone wants it on v13, here it is.

This PR is made of 1 automatic OCA module_migrator PR + 1 manual PR to make test pass on v13 https://github.com/OCA/l10n-brazil/pull/1568/commits/32d3aabcf6c8427183a187e56879662a917256e7

### CODE REVIEW

File: l10n_br_base/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright (C) 2009 - TODAY Renato Lima - Akretion
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    "name": "Brazilian Localization Base",
+    "summary": "Customization of base module for implementations in Brazil.",
+    "category": "Localization",
+    "license": "AGPL-3",
+    "author": "Akretion,Odoo Community Association (OCA)",
```
Comment (renatonlima) [inline l10n_br_base/__manifest__.py:9]: Além do autor seria interessante colocar os mantenedores com a chave "maintainers"

**Review by renatonlima — APPROVED** [review]

**Review by marcos-mendez — APPROVED** [review]
Aqui foi, seria legal já atualizar os postgres tmb?

**Comment by britoederr** [comment]
image

Error when placing invalid CPF/CNPJ in the respective field of the company

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/home/odoo/build/OCA/l10n-brazil/l10n_br_base/models/res_company.py", line 137, in write
    result = super(Company, self).write(values)
  File "/.repo_requirements/odoo/addons/web/models/models.py", line 403, in write
    res = super().write(values)
  File "/.repo_requirements/odoo/odoo/addons/base/models/res_company.py", line 252, in write
    return super(Company, self).write(values)
  File "/.repo_requirements/odoo/odoo/models.py", line 3639, in write
    fields[0].determine_inverse(real_recs)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1154, in determine_inverse
    getattr(records, self.inverse)()
  File "/home/odoo/build/OCA/l10n-brazil/l10n_br_base/models/res_company.py", line 46, in _inverse_cnpj_cpf
    company.partner_id.cnpj_cpf = company.cnpj_cpf
  File "/.repo_requirements/odoo/odoo/fields.py", line 1121, in __set__
    records.write({self.name: write_value})
  File "/.repo_requirements/odoo/addons/base_address_extended/models/base_address_extended.py", line 153, in write
    res = super(Partner, self).write(vals)
  File "/.repo_requirements/odoo/odoo/addons/base/models/res_partner.py", line 528, in write
    result = result and super(Partner, self).write(vals)
  File "/.repo_requirements/odoo/odoo/models.py", line 3634, in write
    real_recs._validate_fields(vals, inverse_fields)
  File "/.repo_requirements/odoo/odoo/models.py", line 1206, in _validate_fields
    check(self)
  File "/home/odoo/build/OCA/l10n-brazil/l10n_br_base/models/res_partner.py", line 113, in _check_cnpj_cpf
    raise ValidationError(_("{} Invalid!").format(document))
odoo.exceptions.ValidationError: ('CNPJ Invalid!', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/.repo_requirements/odoo/addons/web/controllers/main.py", line 1339, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/.repo_requirements/odoo/addons/web/controllers/main.py", line 1331, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 387, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 374, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/build/OCA/l10n-brazil/l10n_br_base/models/res_company.py", line 142, in write
    raise Exception
Exception

File: l10n_br_base/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        return res
+
+    def write(self, values):
+        try:
+            result = super(Company, self).write(values)
+        except Exception:
+            if not config["without_demo"] and values.get("currency_id"):
+                result = models.Model.write(self, values)
+            else:
+                raise Exception
```
Comment (britoederr) [inline l10n_br_base/models/res_company.py]: Não seria interessante usar uma forma mais elegante de exibir o erro ao usuário?
Algo como: raise exceptions.UserError('Há algum erro no cadastro. Confira todos os campos')

Comment (rvalyi) [reply l10n_br_base/models/res_company.py]: hehe vamos ter mais gente contribuindo assim que a gente migrar, esse é o objetivo tb. Na vdd esse erro deveria apenas propagar o erro original. O que vc acha de https://github.com/OCA/l10n-brazil/pull/1573 ?

File: l10n_br_base/__manifest__.py (hunk 1 above)
Comment (rvalyi) [reply l10n_br_base/__manifest__.py:9]: https://github.com/OCA/l10n-brazil/pull/1573

**Comment by rvalyi** [comment]
> Aqui foi, seria legal já atualizar os postgres tmb?

O OCA/l10n-brazil roda com qualquer versão Postgres que o Odoo roda. Nos testes a gente não especifica versão não, deixa isso com o padrão da OCA e é melhor assim.

File: l10n_br_base/models/res_company.py (hunk 2 above)
Comment (britoederr) [reply l10n_br_base/models/res_company.py]: @rvalyi  Show! bem melhor pro usuário!

Comment (rvalyi) [reply l10n_br_base/models/res_company.py]: Sim valeu por ter detetado!

**Comment by rvalyi** [comment]
@renatonlima @britoederr fixes from 12.0 included via cherry-pick. How about e review?

**Review by britoederr — APPROVED** [review]

**Pull request merged**